### PR TITLE
feat: add C implementation for `math/base/special/gammaln`

### LINKED ISSUE #1988: [RFC]: add C implementation for `math/base/special/gammaln`
### Description

This RFC proposes adding native C implementation for `@stdlib/math/base/special/gammaln`
```javascript
double stdlib_base_gammaln ( const double x );
```

### Related Issues

Issue Tracker: #649 

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by gunjjoshi**
Resolves #1988.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/gammaln`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/gammaln).

```c
double stdlib_base_gammaln( const double x );
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649.

## Questions

> Any questions for reviewers of this pull request?

Shall I change both `test.js` and `test.native.js` to use tolerance of the form `1.0 * EPS` here? We have a high amount of tolerance required here as well, just like [`math/base/special/gamma`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/gamma/test/test.js#L118).

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by kgryte** [comment]
> Shall I change both test.js and test.native.js to use tolerance of the form 1.0 * EPS here? 

Yes, I think that would be good.

File: lib/node_modules/@stdlib/math/base/special/gammaln/test/test.js [hunk 1]
```diff
@@ -87,31 +87,49 @@ tape( 'the function returns `-ln(x)` for very small x', function test( t ) {
...
+	x = data2;
+	expected = expected2;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = gammaln( x[ i ] );
+		if ( y === expected[ i ] ) {
+			t.equal( y, expected[ i ], 'x: '+x[i]+'. Expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[ i ] );
+			tol = 1.1e4 * EPS * abs( expected[ i ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/gammaln/test/test.js:130]: @Planeshifter This is also a rather large tolerance. I wonder if the issue is R. Would be good to compare against Julia and NumPy/SciPy. Or more perhaps even more ideally, we'd compare against an arbitrary precision implementation.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/gammaln/test/test.js:130]: For `gamma`, we use Cephes. In this case, looks like we use `fdlibm`.

File: lib/node_modules/@stdlib/math/base/special/gammaln/src/main.c [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*
+* ## Notice
+*
+* The following copyright, license, and long comment were part of the original implementation available as part of [FreeBSD]{@link https://svnweb.freebsd.org/base/release/9.3.0/lib/msun/src/e_lgamma_r.c}. The implementation follows the original, but has been modified for JavaScript.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/gammaln/src/main.c]: @gunjjoshi Did you happen to check this implementation against the latest implementation in FreeBSD?

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/gammaln/src/main.c]: I checked the [latest implementation](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_lgamma_r.c?revision=367086&view=markup), looks like there are some changes in how they're calculating `sin_pi`. For `gammaln`, even in the [older version](https://svnweb.freebsd.org/base/release/9.3.0/lib/msun/src/e_lgamma_r.c?revision=268523&view=markup), they have used `EXTRACT_WORDS`, and are going by a different manner in the beginning, unlike our `javascript` implementation. We haven't used this way of extracting words and shifting them. This is same in both the versions.

**Comment by gunjjoshi** [comment]
I've made changes in both `C` and `javascript` implementations, according to the latest version, and have updated the links.

File: lib/node_modules/@stdlib/math/base/special/gammaln/test/test.js (hunk 1 above)
Comment (Planeshifter) [reply lib/node_modules/@stdlib/math/base/special/gammaln/test/test.js:130]: Investigated and found that Julia is way closer to our implementation (`8.5 * EPS` to get tests to pass.)

**Review by Planeshifter — APPROVED** [review]
Let's land this PR.

I investigated and found that Julia is way closer to our implementation (`8.5 * EPS` to get tests to pass), so that the large tolerances are due to R using a different algorithm. We should add Julia test fixtures and use them in the unit tests, but that can be done in a follow-up PR.

**Review by Planeshifter — APPROVED** [review]

**Comment by gunjjoshi** [comment]
Thanks for checking this @Planeshifter! I'll make the changes in a follow-up PR.

**Comment by Planeshifter** [comment]
@gunjjoshi Thank you!

It may also be useful to add Julia fixtures to `gamma`, as well. I propose we leave the `r` fixtures directory intact and just add another one for Julia, and update the test files to use the Julia ones.

**Pull request merged**